### PR TITLE
Change Timer to use ConcurrentQueue to send Timer events

### PR DESCRIPTION
**Pull request by jaedan**
The timer thread passes messages indicating that timers have expired
to the main thread via a queue. This queue was previously guarded by
a lock which greatly hurt performance. Instead, use a ConcurrentQueue
which is implemented using atomic compare and swap operations, or on
very new platforms that support it, TSX transactions. This should
provide a reasonable performance improvement.

The queue itself is single consumer, single producer, so if a suitable
implementation of a ring was written, no locks or atomics are strictly
required. This optimization is left for the future.

### CODE REVIEW

File: Server/Timer.cs [hunk 1]
```diff
@@ -364,39 +364,34 @@ public void TimerMain()
...
+			int count = 0;
 
-				int index = 0;
+			while (count++ < m_BreakCount) {
 
-				while ( index < m_BreakCount && m_Queue.Count != 0 )
-				{
-					Timer t = m_Queue.Dequeue();
-					TimerProfile prof = t.GetProfile();
+				if (!m_Queue.TryDequeue(out Timer t)) {
```
Comment (kamronbatman) [inline Server/Timer.cs]: I think the CI complained about this line.  Maybe it should be:
`if (!m_Queue.TryDequeue(out t)) {`

**Review by kamronbatman — CHANGES_REQUESTED** [review]
I really like this (and anything that improves performance). Regarding the future optimization, I haven't used that before. I would love to see an example of that, even if it uses a newer .NET standard.

Also if you have other optimization ideas, I would love to see them!

**Comment by benlwalker** [comment]
A lock is most likely implemented as an atomic compare and swap (CAS), so acquiring a lock or using this would seem to be the same cost at first. However, this queue implementation does the CAS on a different variable for enqueue than it does for dequeue, so it eliminates the contention on the lock that exists now. It should be a big win, but I don't have a server with sufficient scale to measure it.

**Comment by jaedan** [comment]
A lock is most likely implemented as an atomic compare and swap (CAS), so acquiring a lock or using this would seem to be the same cost at first. However, this queue implementation does the CAS on a different variable for enqueue than it does for dequeue, so it eliminates the contention on the lock that exists now. It should be a big win, but I don't have a server with sufficient scale to measure it.

**Comment by kamronbatman** [comment]
@msturgill if I am understanding the code properly, the Enqueue is done on the TimerThread, and the Dequeue (Slice) is done on the main thread.

According to https://docs.microsoft.com/en-us/dotnet/standard/collections/thread-safe/when-to-use-a-thread-safe-collection
> In this scenario, ConcurrentQueue<T> performs best when one dedicated thread is queuing and one dedicated thread is de-queuing.

This would mean that in most scenarios, a ConcurrentQueue would perform better, especially at scale.

Also regarding C# 7, there is quite a bit to discussion considering Mono since it ships with Roslyn now. I'll start a separate issue where we can discuss this.

**Pull request closed without merging**